dask/array/core.py

```diff
@@ -716,6 +716,70 @@ def compute(self, **kwargs):
         result, = compute(self, **kwargs)
         return result
 
+    def cache(self, store=None, **kwargs):
+        """ Evaluate and cache array
+
+        This triggers evaluation and store the result in either
+
+        1.  An ndarray object supporting setitem (see da.store)
+        2.  A MutableMapping like a dict or chest
+
+        It then returns a new dask array that points to this store.
+        This returns a semantically equivalent dask array.
+
+        >>> import dask.array as da
+        >>> x = da.arange(5, chunks=2)
+        >>> y = 2*x + 1
+        >>> z = y.cache()  # triggers computation
+
+        >>> y.compute()  # Does entire computation
+        array([1, 3, 5, 7, 9])
+
+        >>> z.compute()  # Just pulls from store
+        array([1, 3, 5, 7, 9])
+
+        You might base a cache off of an array like a numpy array or
+        h5py.Dataset.
+
+        >>> cache = np.empty(5, dtype=x.dtype)
+        >>> z = y.cache(store=cache)
+        >>> cache
+        array([1, 3, 5, 7, 9])
+
+        Or one might use a MutableMapping like a dict or chest
+
+        >>> cache = dict()
+        >>> z = y.cache(store=cache)
+        >>> cache  # doctest: +SKIP
+        {('x', 0): array([1, 3]),
+         ('x', 1): array([5, 7]),
+         ('x', 2): array([9])}
+        """
+        if store is not None and hasattr(store, 'shape'):
+            self.store(store)
+            return from_array(store, chunks=self.chunks)
+        if store is None:
+            try:
+                from chest import Chest
+                store = Chest()
+            except ImportError:
+                if self.nbytes <= 1e9:
+                    store = dict()
+                else:
+                    raise ValueError("No out-of-core storage found."
+                        "Either:\n"
+                        "1. Install ``chest``, an out-of-core dictionary\n"
+                        "2. Provide an on-disk array like an h5py.Dataset") # pragma: no cover
+        if isinstance(store, MutableMapping):
+            name = next(names)
+            dsk = dict(((name, k[1:]), (operator.setitem, store, (tuple, list(k)), k))
```

Comment:
This stops `k` from being evaluated.  We don't want its value, we actually want the tuple itself.

`k` is a tuple like `('x', 0, 1)`.   This value is a key in our graph and so, unless we do something about it, `k` will be replaced with the full array rather than the tuple as we want.  `(tuple, list(k))` turns it into a list while in the graph but executes it back into a tuple when it's time to execute `setitem`.

That was complex, let me try again.

We want to do this

    store[('x', 0, 1)] = some_numpy_array

where `some_numpy_array` is the result corresponding to computing `('x', 0, 1)` from the graph.